apps/demo-app/pages/api/check-signature.ts

```diff
@@ -0,0 +1,205 @@
+
+  return verifyADR36Amino(
+    "xion",
+    address,
+    messageString,
+    pubKeyUint8Array,
+    uint8Signature,
+  );
+}
+
+/**
+ * Verifies the Xion signature and grants for a given address, session address, public key,
+ * message string, and signature.
+ *
+ * @param address - The address to verify the grants for.
+ * @param sessionAddress - The session address to verify the grants for.
+ * @param pubKey - The public key associated with the session address base64 encoded.
+ * @param messageString - The message string to verify the signature against.
+ * @param signature - The signature to verify.
+ *
+ * @returns Promise<boolean> - A promise that resolves to true if the Xion signature and grants are valid,
+ *                             or false otherwise.
+ */
+export async function verifyXionSignatureAndGrants(
+  address: string,
+  sessionAddress: string,
+  pubKey: string,
+  messageString: string,
+  signature: string,
+): Promise<boolean> {
+  const isValid = verifyXionSignature(
+    sessionAddress,
+    pubKey,
+    messageString,
+    signature,
+  );
+  if (!isValid) {
+    return false;
+  }
+
+  return verifyGrants(address, sessionAddress);
+}
+
+/**
+ * Verifies if grants have been given by a granter to a grantee.
+ *
+ * @param grantee - The address of the granter.
+ * @param granter - The address of the grantee.
+ *
+ * @returns Promise<boolean> - A promise that resolves to a boolean indicating whether ANY grants have been given.
+ */
+export async function verifyGrants(
+  granter: string,
+  grantee: string,
+): Promise<boolean> {
+  const baseUrl = `${testnetChainInfo.rest}/cosmos/authz/v1beta1/grants`;
+  const url = new URL(baseUrl);
+  const params = new URLSearchParams({
+    grantee,
+    granter,
+  });
+  url.search = params.toString();
+  const data = await fetch(url, {
+    cache: "no-store",
+  })
+    .then((response): Promise<QueryGrantsResponse> => response.json())
+    .catch((err) => {
+      console.error("Could not fetch grants info", err);
+    });
+
+  return Boolean(data && data.grants.length > 0);
+}
+
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse,
+) {
+  const { query, method } = req;
+
+  if (method === "GET") {
+    const {
+      userSessionAddress,
+      userSessionPubKey,
+      metaAccountAddress,
+      message,
+      signature,
+    } = query;
+
+    const errors: string[] = [];
+    if (!userSessionAddress) {
+      errors.push("userSessionAddress is required");
+    }
+
+    if (!userSessionPubKey) {
+      errors.push("userSessionPubKey is required");
+    }
+
+    if (!metaAccountAddress && typeof metaAccountAddress === "string") {
+      errors.push("itemId is required");
+    }
+
+    if (!message && typeof message === "string") {
+      errors.push("message is required");
+    }
+
+    if (!signature && typeof signature === "string") {
+      errors.push("signature is required");
+    }
```

Comment:
Could maybe condense these and colocate the error messages so it's easier to tell what the issue is at a quick glance. Not an issue worth not approving, but a thought.

```typescript
const requiredParams = {
  userSessionAddress: "userSessionAddress is required",
  userSessionPubKey: "userSessionPubKey is required",
  metaAccountAddress: "metaAccountAddress is required",
  message: "message is required",
  signature: "signature is required",
};

for (const [param, errorMsg] of Object.entries(requiredParams)) {
  if (!query[param] || !isString(query[param])) {
    errors.push(errorMsg);
  }
}
```

Reply:
That is clever!

Reply:
Sadly, typescript figure out the types like this correctly :(